BranchMetrics/android-branch-deep-linking-attribution
# SDK-1924 -- [Android] Enable automated unit and instrumented tests via GHA

## Reference
SDK-1924 -- [Android] Enable automated unit and instrumented tests via GHA

## Description
Presently unit and instrumented tests are manually run as part of our contribution process. This PR automatically runs tests per each push using GitHub Actions.

The workflow consists of two jobs parallelizing the unit and instrumented tests, both of which are also parallelized to two other jobs by API level, currently targeting the minimum API we support as well as the current.

The workflow itself defines and boots emulator on demand using the runner's preinstalled Android platform tools. Because project's gradle is configured to use Java 11, we must also set up Java prior to running gradle commands.

Additionally, added gradle task to format and display test results in logs.

Requirements:
- Runs unit tests
- Runs instrumented tests
- With minimum API **21**
- And with current API **33**
- Uses only verified actions.
- Results are accessible in logs and artifact
  - With reproducible apk, structured result report artifacts

## Testing Instructions
Check out the branch, push a trivial commit to trigger the workflow. Inspect on your PR the tests are queued. executed, and resulted as part of the checks on the PR summary page.

With passing tests 
https://github.com/BranchMetrics/android-branch-deep-linking-attribution/actions/runs/4792510537

With failing tests
https://github.com/BranchMetrics/android-branch-deep-linking-attribution/actions/runs/4792331072

## Risk Assessment [`LOW`]



- [✅] I, the PR creator, have tested — integration, unit, or otherwise — this code.

## Reviewer Checklist (To be checked off by the reviewer only)

- Correctness & Style

cc @BranchMetrics/saas-sdk-devs for visibility.
